apps/verticalhome/js/pinned_apps_navigation.js

```diff
@@ -27,260 +27,354 @@
-            }
-
-            this.selectedElemIndex--;
+          if(this.rearrangeMode){
+           this.moveItemDown();
+           this.refresh();
+          }
+          else{
+          this.selectedElemIndex = this.currentIndex() || 0;
+
+          if(this.selectedElemIndex<=0)
+          {
+            this.selectedElemIndex = 0;
+            this.elements[0].classList.add('selected');
+            needClone = false;
+          }
 
-            var removed = this.elements.splice(0, 1);
+          currSelected = this.elements[this.selectedElemIndex];
+          nextSelected = this.elements[this.selectedElemIndex+1];
 
-            var cloned = removed[0].cloneNode(true);
+          nextSelected.classList.add('selected');
 
-            removed[0].classList.add('removed');
+          if(needClone){
+            itm = this.elements[0];
+            cln = itm.cloneNode(true);
 
-            this.elemList.appendChild(cloned);
+            this.elemList.appendChild(cln);
+            itm.parentNode.removeChild(itm);
+          }
 
-            if (app.pinnedAppsManager.items[+cloned.dataset.index]){
-              app.pinnedAppsManager.
-                  items[+cloned.dataset.index].
-                  refreshDomElem(cloned);
-            }
+          if( currSelected.classList.contains('selected')){
+           currSelected.classList.remove('selected');
+          }
+          if( currSelected.classList.contains('middle')){
+          currSelected.classList.remove('middle');
+          }
 
-            this.elements[this.elements.length] = cloned;
+          this.refresh();
+          this.scrollNum();
 
-            this.selectedElemIndex++;
-            this.elements[this.selectedElemIndex].classList.add('selected');
+          }
+          break;
 
-            this.elemList.style.top = this.elemList.dataset.scrolldown;
-          } else {
-            this.selectedElemIndex++;
-            this.elements[this.selectedElemIndex].classList.add('selected');
+        case 'Accept':
+          var ElsSelectedIndex = this.elements[this.selectedElemIndex];
 
-            topStyle = parseFloat(this.elemList.style.top);
-            this.elemList.style.top = topStyle - 6 + 'rem';
-            this.elemList.dataset.scrolldown = this.elemList.style.top;
+          if (this.personalizeMode){
+            if(ElsSelectedIndex.getAttribute('id') == 'addFromMoreApps'){
+                app.enterMoreAppsPersonalise();
+             }
           }
+          else{
 
-          this.prevVerticalKey = 'down';
+            if (ElsSelectedIndex.getAttribute('id') == 'moreApps') {
+              app.showMoreApps();
+            } else {
+              app.getAppByURL(ElsSelectedIndex.dataset.manifesturl).launch();
+            }
 
-          if (this.selectedElemIndex == 1) {
-            this.elemList.style.top = '0rem';
           }
 
           break;
+        case 'q': // enter to personalize mode
```

Comment:
If I understand correctly you are also integrating support of softkeys in this patch, so why do we need 'magic' keys bindings? If it's like a backup plan until softkeys are not integrated, please live appropriate comment in the code, like: ``//TODO: remove these bindings when....``